app/src/main/java/org/simple/clinic/scheduleappointment/ScheduleAppointmentSheetController.kt

```diff
@@ -0,0 +1,99 @@
+package org.simple.clinic.scheduleappointment
+
+import io.reactivex.Observable
+import io.reactivex.ObservableTransformer
+import io.reactivex.rxkotlin.ofType
+import io.reactivex.rxkotlin.withLatestFrom
+import org.simple.clinic.ReportAnalyticsEvents
+import org.simple.clinic.overdue.AppointmentRepository
+import org.simple.clinic.widgets.UiEvent
+import org.threeten.bp.LocalDate
+import org.threeten.bp.ZoneOffset.UTC
+import org.threeten.bp.temporal.ChronoUnit
+import javax.inject.Inject
+
+typealias Ui = ScheduleAppointmentSheet
+typealias UiChange = (Ui) -> Unit
+
+class ScheduleAppointmentSheetController @Inject constructor(
+    private val repository: AppointmentRepository
+) : ObservableTransformer<UiEvent, UiChange> {
+
+  override fun apply(upstream: Observable<UiEvent>): Observable<UiChange> {
+    val replayedEvents = upstream.compose(ReportAnalyticsEvents()).replay().refCount()
+
+    return Observable.mergeArray(
+        sheetCreates(replayedEvents),
+        dateIncrements(replayedEvents),
+        dateDecrements(replayedEvents),
+        schedulingSkips(replayedEvents),
+        scheduleCreates(replayedEvents))
+  }
+
+  private fun sheetCreates(events: Observable<UiEvent>): Observable<UiChange> {
+    return events.ofType<SheetCreated>()
+        .map { { ui: Ui -> ui.updateDisplayedDate(it.initialIndex) } }
+  }
+
+  private fun dateIncrements(events: Observable<UiEvent>): Observable<UiChange> {
+    val lastPositionStream = events.ofType<IncrementAppointmentDate>()
+        .filter { (current, size) -> isSecondLastItem(current, size) }
+        .map { { ui: Ui -> ui.enableIncrementButton(false) } }
+
+    val notLastPositionStream = events.ofType<IncrementAppointmentDate>()
+        .filter { (current, size) -> isNotLastItem(current, size) }
+
+    val updateDateStream = notLastPositionStream
+        .map { { ui: Ui -> ui.updateDisplayedDate(newIndex = it.currentIndex.inc()) } }
+
+    val enableButtonStream = notLastPositionStream
+        .distinctUntilChanged()
+        .map { { ui: Ui -> ui.enableIncrementAndDecrementButtons() } }
+
+    return Observable.merge(lastPositionStream, updateDateStream, enableButtonStream)
+  }
+
+  private fun isSecondLastItem(current: Int, size: Int) = current == size - 2
+
+  private fun isNotLastItem(current: Int, size: Int) = current != size - 1
```

Comment:
Considering that both `isSecondLastItem()` and `isNotLastItem()` aren't used by any other functions, they can be inlined inside `dateIncrements()` as lambdas.